forch/device_report_client.py

```diff
@@ -56,27 +59,42 @@ def _connect(self, mac, vlan, assigned):
         session_params.device_mac = mac
         session_params.device_vlan = vlan
         session_params.assigned_vlan = assigned
-        progresses = self._stub.StartSession(session_params)
-        thread = threading.Thread(target=lambda: self._process_progress(mac, progresses))
+        session = self._stub.StartSession(session_params)
+        thread = threading.Thread(target=lambda: self._process_progress(mac, session))
         thread.start()
-        return thread
+        return session
+
+    def disconnect(self, mac):
+        with self._lock:
+            session = self._mac_sessions.get(mac)
+            if session:
+                session.cancel()
+                self._mac_sessions.pop(mac)
+                self._logger.info('Device %s disconnected', mac)
+            else:
+                self._logger.warning('Disconnecting unconnected device %s', mac)
```

Comment:
NIT: Should this log message be more on the lines of 'Can't disconnect unconnected device' or 'Device not connected' since the method doesn't actually do anything if it is disconnected?

Reply:
Yup -- updated.